python/lsst/pipe/tasks/setPrimaryFlags.py

```diff
@@ -20,12 +20,164 @@
+        True for each source that has a centroid in the same
+        tract as `tractInfo`.
+    """
+    # Mark sources in the inner region of a tract
+    tractId = tractInfo.getId()
+    isTractInner = np.array([skyMap.findTract(s.getCoord()).getId() == tractId for s in sources])
+    return isTractInner
+
+
+def getPseudoSources(sources, pseudoFilterList, schema, log):
+    """Get a flag that marks pseudo sources
+
+    Some categories of sources, for example sky objects,
+    are not really detected sources and should not be considered primary
+    sources.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog.
+    pseudoFilterList: `list` of `str`
+        Names of filters which should never be primary
+
+    Returns
+    -------
+    isPseudo: `bool` array
+        True for each source that is a pseudo source.
+        Note: to remove pseudo sources use `~isPseudo`.
+    """
+    # Filter out sources that should never be primary
+    isPseudo = np.zeros(len(sources), dtype=bool)
+    for filt in pseudoFilterList:
+        try:
+            pseudoFilterKey = schema.find("merge_peak_%s" % filt).getKey()
+            isPseudo |= sources[pseudoFilterKey]
+        except KeyError:
+            log.warn("merge_peak is not set for pseudo-filter %s" % filt)
+    return isPseudo
+
+
+def getDeblendPrimaryFlags(sources, nChildKey=None):
+    """Get both types of primary sources flagged by the deblender
+
+    scarlet is different than meas_deblender in that it is not
+    (necessarily) flux conserving. For consistency in scarlet,
+    all of the parents with only a single child (isolated sources)
+    need to be deblended. This creates a question: which type
+    of isolated source should we make measurements on, the
+    undeblended "parent" or the deblended child?
+    For that reason we distinguish between a SimpleLeaf,
+    which is a source that has no children and uses the
+    isolated parents, and a ModelLeaf, which uses
+    the scarlet models for both isolated and blended sources.
+    In the case of meas_deblender, a ModelLeaf is the same thing
+    as a blended source, and a SimpleModel has the same definition.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog that has already been deblended using
+        either meas_extensions_scarlet or meas_deblender.
+    nChildKey: `str`
+        Name of the column in `sources` that contains the
+        number of deblended children for each source.
+
+    Returns
+    -------
+    isSimpleLeaf: `bool` array
+        True for each source that is a `SimpleLeaf` as defined above.
+    isModelLeaf: `bool` array
+        True for each sources that is a `ModelLeaf` as defined above.
+    """
+    if nChildKey is None:
+        nChildKey = "deblend_nChild"
+    nChild = sources[nChildKey]
+    parent = sources["parent"]
+    isModelLeaf = (parent != 0) & (nChild == 0)
```

Comment:
Doesn't this explicitly exclude all truly isolated sources (i.e. with `parent == 0 & nChild == 0`)?

Reply:
No, because there are no `parent==0 & nChild == 0` sources in the scarlet outputs. `nChild` describes the number of children in the catalog deblended from a parent, and because scarlet will `deblend` all of the isolated sources, all of the isolated parents have `nChild == 1`. Maybe that's what you were missing? If you can think of a way for me to make that more clear in the docs let me know.